packages/backend/src/helpers/actions/handle-failed-step-and-throw.ts

```diff
@@ -0,0 +1,193 @@
+import type { IActionJobData } from '@plumber/types'
+
+import { type JobPro, WorkerPro } from '@taskforcesh/bullmq-pro'
+import { UnrecoverableError } from '@taskforcesh/bullmq-pro'
+import { type Span } from 'dd-trace'
+import get from 'lodash.get'
+
+import HttpError from '@/errors/http'
+import RetriableError from '@/errors/retriable-error'
+import StepError from '@/errors/step'
+import Execution from '@/models/execution'
+import ExecutionStep from '@/models/execution-step'
+
+import { MAXIMUM_JOB_ATTEMPTS } from '../default-job-configuration'
+
+function handleRetriableError(
+  executionError: RetriableError,
+  context: HandleFailedStepAndThrowParams['context'],
+): never {
+  const { delayType, delayInMs } = executionError
+  const { worker, job, isQueuePausable } = context
+
+  switch (delayType) {
+    case 'queue':
+      if (isQueuePausable) {
+        worker.rateLimit(delayInMs)
```

Comment:
check for attemptsStarted / attemptsMade since attempts are ignored if ratelimited

Reply:
I thought about this again, and I think we should still allow the queue / group jamming to go through. Reason: the API has already rate limited us, but if we throw `UnrecoverableError` here, we will process the next step immediately while still being rate limited.

The right move should be to check for `attemptsStarted` _before_ calling `processAction`. This jams the group / queue so that we don't call the API again while being rate limited, but also fails the step without exposing us to possibility of infinite retries.

This is a riskier change, so I think we can do it after I come back. The tradeoff is that there is a risk we retry _forever_ if we keep getting 429s, but i think this risk is small enough. To mitigate, I added `attemptsStarted` in onError callback as a safeguard; we can set an alarm on that if there's a job that has `attempsStarted` >= 20 times or similar.